lighthouse.go

```diff
@@ -456,30 +461,12 @@ func (lh *LightHouse) Query(ip iputil.VpnIp, f EncWriter) *RemoteList {
 	return nil
 }
 
-// This is asynchronous so no reply should be expected
-func (lh *LightHouse) QueryServer(ip iputil.VpnIp, f EncWriter) {
+// QueryServer is asynchronous so no reply should be expected
+func (lh *LightHouse) QueryServer(ip iputil.VpnIp) {
 	if lh.amLighthouse {
 		return
 	}
-
-	if lh.IsLighthouseIP(ip) {
-		return
-	}
-
-	// Send a query to the lighthouses and hope for the best next time
-	query, err := NewLhQueryByInt(ip).Marshal()
-	if err != nil {
-		lh.l.WithError(err).WithField("vpnIp", ip).Error("Failed to marshal lighthouse query payload")
-		return
-	}
-
-	lighthouses := lh.GetLighthouses()
-	lh.metricTx(NebulaMeta_HostQuery, int64(len(lighthouses)))
-	nb := make([]byte, 12, 12)
-	out := make([]byte, mtu)
-	for n := range lighthouses {
-		f.SendMessageToVpnIp(header.LightHouse, 0, n, query, nb, out)
-	}
+	lh.queryChan <- ip
```

Comment:
The channel consumer in startQueryWorker() calls lh.ifce.SendMessageToVpnIp() calls f.getOrHandshake() calls hm.StartHandshake() calls hm.lightHouse.QueryServer()

This call chain means that the channel consumer may also attempt to write to the channel. If the channel is full, this write operation will block forever, as the only thing that could make room in the channel is blocked waiting for room in the channel.

Reply:
The call stack can only happen if QueryWorker doesn't have a tunnel to some lighthouse. With an early exit in QueryServer for lighthouse IP's, then QueryWorker will never try to write to the channel that it's reading from.